src/main/java/edu/umich/its/spe/SPESummary.java

```diff
@@ -66,24 +72,31 @@ public String toString() {
 		endTime = LocalDateTime.now();
 		StringBuffer result = new StringBuffer();
 
-		 Duration dur = Duration.between(startTime, endTime);
-
-		 //System.out.println("start: "+startTime+" end: "+endTime+" duration: "+DurationFormatUtils.formatDurationHMS(dur.toMillis()));
+		Duration dur = Duration.between(startTime, endTime);
 
+		// make sure there is a printable value even if no course id (when testing).
+		String courseIdString = courseId.toString().length() > 0 ? courseId.toString() : "[none]";
 
+		result.append(LINE_RETURN);
+		result.append(LINE_RETURN);
 		result.append("starting time: ").append(PersistTimestamp.formatTimestamp(startTime)).append(LINE_RETURN);
 		result.append("end time: ").append(PersistTimestamp.formatTimestamp(endTime)).append(LINE_RETURN);
 		result.append("elapsed time: ").append(DurationFormatUtils.formatDurationHMS(dur.toMillis())).append(LINE_RETURN);
+		result.append(LINE_RETURN);
 
 		result.append("storedGradesLastRetrieved: ").append(storedGradesLastRetrieved.toString()).append(LINE_RETURN);
 		result.append("useGradesLastRetrieved: ").append(useGradesLastRetrieved.toString()).append(LINE_RETURN);
 		result.append("updatedGradesLastRetrieved: ").append(updatedGradesLastRetrieved.toString()).append(LINE_RETURN);
+		result.append(LINE_RETURN);
 
-		result.append("courseId: ").append(courseId.toString()).append(LINE_RETURN);
+		result.append("courseId: ").append(courseIdString).append(LINE_RETURN);
+		result.append(LINE_RETURN);
 
 		result.append("users added: ").append(added).append(" errors: ").append(errors).append(LINE_RETURN);
+		result.append(LINE_RETURN);
 
-		users.forEach((u) -> result.append("user: ").append(u.getLeft())
+		users.forEach((u) -> result
+				.append("user: ").append(u.getLeft())
```

Comment:
I feel instead of having appends like this SPE can have a template of the summary and replace the placeholders in the SPESummany class
```
# Template
This is a summary
starting time: <placeholder to replace> 
end time: <placeholder to replace>
etc...
```

Reply:
That's a fine way to do it.  This was the quickest way to do it.  If it needs to be revised I'll look into using a template.